### DISCUSSION TITLE
Need help with: Strange behaviour in NextJS API routing. params is undefined although the route is correct.

### DISCUSSION THREAD

**Discussion by JSulthoni** [vercel/next.js#58995] [Help] [answered]
### Summary

Hello, 👋. I am currently building an app using NextJS and having trouble of connecting my frontend to the API routes.

I have followed the routing guide by NextJS [here](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes) but when I tried to make a request to the route i got error message of:

>  ⨯ TypeError: Cannot destructure property 'id' of 'params' as it is undefined.

This is the relative path to my endpoint:
`.\src\app\api\comments\[id]\route.js`

This is the code i have in my endpoint:
```ts
export const GET = async ({ params }) => {
    const { id } = params; // This should destructure id from params
    console.log(id) // This did not log the expected id


    const session = await getAuthSession();

    if (!session) {
        return new NextResponse(
            JSON.stringify({ message: "Not Authenticated!" }, { status: 401 })
        )
    };

    try {
        const comment = await prisma.comment.findUnique({
            where: { id }
        });
        console.log(`GETTING COMMENT with ID: ${id}`);
        return new NextResponse(JSON.stringify(comment, { status: 200 }));
        
    } catch (error) {
        console.error(error);
        return new NextResponse(JSON.stringify({message: `Failed to GET comment with id of ${id}` }, {status: 500}));
    }
};
```

I did not understand why I got this error. I expected if I make a get request (e.g `http://localhost:3000/api/comments/1`) at least params should be not undefined and i can get the ID from the params, and the console should log 1 from the `console.log(id)` line of code.

Getting further into the problem, within the same route, I have tried a code like this.
```ts
export const GET = ({ params }) => {
   console.log('Params: ', params) // This logs undefined
    return new NextResponse(JSON.stringify({message: params}));
};
```
And by using browser or Insomnia by making request to `http://localhost:3000/api/comments/1` , log says params in undefined.

In my act of despair I have pushed all my current state of code with all the needed env variables [here](https://github.com/JSulthoni/Treadclone) for the discussion. 

note: The repo is currently ongoing another discussion, I'm terribly sorry if the README is not on the topic

### Additional information

```js
Error message:
 ⨯ TypeError: Cannot destructure property 'id' of 'params' as it is undefined.
```

**Comment by icyJoseph** [comment, marked as answer]
Hi,

According to the docs, the first argument passed to the Router Handler, is the Request object.

The second argument is an object with the dynamic params.

```ts
export async function GET(
  request: Request,
  { params }: { params: { slug: string } }
) {
  const slug = params.slug // 'a', 'b', or 'c'
}
```

In your case you can do something like:

```ts
export const GET = (request, { params }) => {
    console.log('Params: ', params)
    return new NextResponse(JSON.stringify({message: params}));
};
```

**Reply by JSulthoni** [reply]
Hi @icyJoseph 

Thankyou very much! That's very helpful. It's all stated [here](https://nextjs.org/docs/app/building-your-application/routing/route-handlers#request-body).

I have modified my code like this and now it works great!
```ts
export const GET = async (req, { params }) => { // <-- This is where the problem was
    const { id } = params;
    console.log(id) // This logs the expected id

// rest of the code....
};
```

**Comment by whoismaruf** [comment]
@icyJoseph I have similar issue where my route is `job/[job_id]/candidate/routes.ts`. I have tried according to the documents, but it still gives me the `undefined` TypeError. Can you please give me any clue?

```ts
import { NextRequest, NextResponse } from "next/server";
import fetcher from "@/utils/fetch-data";

export async function GET(
    req: NextRequest,
    res: NextResponse,
    { params }: { params: { job_id: number } }
) {
    const queryParams = req.nextUrl.searchParams;
    const job_id = params.job_id;
    const page = queryParams.get("page");
    const responses = await fetcher(`jobs/${job_id}/candidate?page=${page}`);
    return Response.json(responses);
}
```

**Reply by icyJoseph** [reply]
In API routes there's no response parameter passed to the function, remove the second argument.

Rather, you create the response within the function. 

Read the documentation to make sure you have the correct types too. I didn't really look  at your code beyond the incorrect response argument.

**Reply by whoismaruf** [reply]
Wow! It's solved after removing the second argument. Thanks @icyJoseph 🥹

**Comment by Wasikur** [comment]
Hi guys please help me out. I am building a dashboard using nextjs but when I try to update the data through postman PUT is working fine it is accessing the id but when I try to GET the data using same id it is saying id is undefined 
"
import { NextResponse } from 'next/server';
import connectDB from '@/utils/db';
import Product from '@/utils/models/product/product';

export async function PUT(request: any, content: any) {
  //extracting the id
  const productId = content.params.id;
  const filter = { _id: productId };

  // takin the input to update
  const payload = await request.json();
  console.log(payload);

  // Connect to the MongoDB database
  await connectDB();

  const result = await Product.findOneAndUpdate(filter, payload);
  console.log(`Product with id: ${productId} updated`);
  return NextResponse.json({ result, success: true });
}

// Define the GET request handler to fetch a product by its ID

export async function GET(request: any) {
  if (!request.params) {
    return NextResponse.json({ error: 'Missing product ID' }, { status: 400 });
  }
  //extracting the id
  const productId = request.params.id;
  console.log(request.params);
  const record = { _id: productId };
  console.log(productId);

  // Connect to the MongoDB database
  await connectDB();

  const data = await Product.findOne(record);
  console.log(`Product with id: ${productId} fetched`);
  return NextResponse.json({ data, success: true });
}

"

**Reply by icyJoseph** [reply]
Look at your PUT function signature, it has two arguments, and you read the params from the second, content.

Do the same in your GET function.
